commit c61623144fdc6459de2ebb6aaa0740ca02f5d689 (refs/changes/36/3636/3)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-03-14T22:01:52-07:00 (1 year, 7 months ago)
    
    OS-6753 HVM disk quotas insufficient on larger zvols
    Reviewed by: Trent Mick <trentm@gmail.com>
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index d5e4f00e..6c0f3463 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -1246,8 +1246,19 @@ exports.validate = function (brand, action, payload, options, callback)
     });
 };
 
-function setQuota(dataset, quota, log, callback)
+function setQuota(opts, callback)
 {
+    assert.object(opts, 'opts');
+    assert.string(opts.brand, 'opts.brand');
+    assert.string(opts.dataset, 'opts.dataset');
+    assert.object(opts.log, 'opts.log');
+    assert(opts.hasOwnProperty('quota'), 'opts.quota');
+
+    var brand = opts.brand;
+    var dataset = opts.dataset;
+    var log = opts.log;
+    var quota = opts.quota;
+    var quotaKey = 'quota';
     var newval;
 
     assert(log, 'no logger passed to setQuota()');
@@ -1263,7 +1274,14 @@ function setQuota(dataset, quota, log, callback)
         newval = quota.toString() + 'g';
     }
 
-    zfs(['set', 'quota=' + newval, dataset], log, function (err, fds) {
+    if (brand === 'bhyve') {
+        // For bhyve we want to use refquota so that the quota only applies to
+        // the zonepath dataset, and not the zvol 'disks' that are underneath as
+        // zones/<uuid>/diskX
+        quotaKey = 'refquota';
+    }
+
+    zfs(['set', quotaKey + '=' + newval, dataset], log, function (err, fds) {
         if (err) {
             log.error('setQuota() cmd failed: ' + fds.stderr);
             callback(new Error(rtrim(fds.stderr)));
@@ -1959,15 +1977,19 @@ function destroyVolume(volume, log, callback)
 }
 
 // create a new zvol for a VM
-function createVolume(volume, log, callback)
+function createVolume(volume, opts, callback)
 {
+    assert.object(opts, 'opts');
+    assert.string(opts.brand, 'opts.brand');
+    assert.object(opts.log, 'opts.log');
+
+    var brand = opts.brand;
+    var log = opts.log;
     var refreserv;
     var size;
     var snapshot;
     var tracers_obj;
 
-    assert(log, 'no logger passed for createVolume()');
-
     if (process.env.EXPERIMENTAL_VMJS_TRACING) {
         tracers_obj = traceUntilCallback('create-volume', log, callback);
         callback = tracers_obj.callback;
@@ -1988,9 +2010,25 @@ function createVolume(volume, log, callback)
 
     if (volume.hasOwnProperty('refreservation')) {
         refreserv = volume.refreservation;
-    } else {
-        log.debug('defaulting to refreservation = ' + size);
+    } else if (volume.hasOwnProperty('image_uuid')) {
+        // We can't set refreserv > volsize when making a clone, so set it
+        // to volsize instead.
+        refreserv = size;
+        log.debug({
+            size: size
+        }, 'refreservation unset, setting refreserv=size for cloned disk');
+    } else if (brand === 'kvm') {
+        // For KVM we keep the pre-OS-6753 for now to not break compatibility
+        // but could change this to match bhyve (default case below) in the
+        // future.
         refreserv = size;
+        log.debug({
+            size: size
+        }, 'refreservation unset for kvm, setting refreserv=size for new disk');
+    } else {
+        // We default to letting SmartOS set refreservation for bhyve: OS-6753
+        log.debug('refreservation unset, letting system determine');
+        refreserv = 'auto';
     }
 
     async.series([
@@ -2042,7 +2080,9 @@ function createVolume(volume, log, callback)
                     args.push('-o', 'volblocksize='
                         + volume.block_size);
                 }
-                args.push('-o', 'refreservation=' + refreserv + 'M');
+                if (refreserv !== 'auto') {
+                    args.push('-o', 'refreservation=' + refreserv + 'M');
+                }
                 args.push(snapshot, target);
                 zfs(args, log, function (e) {
                     if (e) {
@@ -2064,8 +2104,10 @@ function createVolume(volume, log, callback)
                     args.push('-o', 'volblocksize='
                         + volume.block_size);
                 }
-                args.push('-o', 'refreservation=' + refreserv + 'M', '-V',
-                    size + 'M', target);
+                if (refreserv !== 'auto') {
+                    args.push('-o', 'refreservation=' + refreserv + 'M');
+                }
+                args.push('-V', size + 'M', target);
                 zfs(args, log, function (err, fds) {
                     if (err) {
                         cb(err);
@@ -2646,7 +2688,10 @@ function createVolumes(payload, log, callback)
     }
 
     function _loggedCreateVolume(volume, cb) {
-        return createVolume(volume, log, cb);
+        return createVolume(volume, {
+            brand: payload.brand,
+            log: log
+        }, cb);
     }
 
     function _loggedDeleteVolume(volume, cb) {
@@ -3037,7 +3082,7 @@ function createVM(payload, log, callback)
             //
             //  zones/<uuid>/disk0
             //
-            args = ['create', '-o', 'quota=' + payload.quota + 'G',
+            args = ['create', '-o', 'refquota=' + payload.quota + 'G',
                 payload.zpool + '/' + payload.uuid];
 
             zfs(args, log, function _onZfsCreate(err) {
@@ -4549,6 +4594,14 @@ function checkPayloadProperties(payload, vmobj, log, callback)
             if (payload.update_disks.hasOwnProperty(disk)) {
                 zvol = payload.update_disks[disk];
 
+                // Disks cannot have refreservation=auto on update.
+                // Only create.
+                if (zvol.refreservation === 'auto') {
+                    callback(new Error('cannot change refreservation to '
+                        + 'auto on existing disk'));
+                    return;
+                }
+
                 if (zvol.hasOwnProperty('compression')) {
                     if (VM.COMPRESSION_TYPES.indexOf(zvol.compression) === -1) {
                         callback(new Error('invalid compression type for '
@@ -9860,50 +9913,6 @@ exports.create = function (payload, options, callback)
                 }
                 cb(err);
             });
-        }, function _fixByveQuota(cb) {
-            var disk;
-            var disks;
-            var idx;
-            var oldQuota;
-
-            // For bhyve, we set quota to quota + the sum of the disk sizes.
-            // This is because here (unlike KVM) the volumes are sub-datasets.
-            // It has to happen after normalizePayload since that will have set
-            // the image_size if the original payload failed to include one.
-            if (payload.brand !== 'bhyve') {
-                cb();
-                return;
-            }
-
-            if (payload.add_disks) {
-                disks = payload.add_disks;
-            } else if (payload.disks) {
-                disks = payload.disks;
-            }
-
-            // These should already have been "normalized"/validated above, and
-            // for bhyve a VM doesn't make sense without disks.
-            assert.number(payload.quota, 'payload.quota');
-            assert.arrayOfObject(disks, 'disks');
-
-            oldQuota = payload.quota;
-            for (idx = 0; idx < disks.length; idx++) {
-                disk = disks[idx];
-
-                if (disk.image_size) {
-                    // image_size is MiB, quota is GiB
-                    payload.quota += Math.ceil(Number(disk.image_size) / 1024);
-                } else if (disk.size) {
-                    payload.quota += Math.ceil(Number(disk.size) / 1024);
-                }
-            }
-
-            log.debug({
-                newQuota: payload.quota,
-                oldQuota: oldQuota
-            }, 'Updated bhyve VM quota to include disks');
-
-            cb();
         }, function (cb) {
             checkDatasetProvisionable(payload, log, function (provisionable) {
                 if (!provisionable) {
@@ -12929,9 +12938,12 @@ function applyUpdates(oldobj, newobj, payload, log, callback)
             if (payload.hasOwnProperty('quota')
                 && (Number(payload.quota) !== Number(oldobj.quota))) {
 
-                setQuota(newobj.zfs_filesystem, payload.quota, log,
-                    function (err) {
-
+                setQuota({
+                    brand: oldobj.brand,
+                    dataset: newobj.zfs_filesystem,
+                    log: log,
+                    quota: payload.quota
+                }, function _onSetQuota(err) {
                     if (!err) {
                         changed = true;
                     }
@@ -13449,7 +13461,7 @@ exports.update = function (uuid, payload, options, callback)
             }
 
             // add the bits of payload createVolumes() needs.
-            p = {'add_disks': disks};
+            p = {add_disks: disks, brand: vmobj.brand};
             p.uuid = uuid;
             if (vmobj.hasOwnProperty('zpool')) {
                 p.zpool = vmobj.zpool;
diff --git a/src/vm/node_modules/proptable.js b/src/vm/node_modules/proptable.js
index 7b3b134c..4092fcd0 100644
--- a/src/vm/node_modules/proptable.js
+++ b/src/vm/node_modules/proptable.js
@@ -35,11 +35,13 @@
  *
  */
 
-// values used in feature table for bhyve
 // This can likely be lowered once we use large pages.
 var BHYVE_MEM_OVERHEAD = 1024 + 256;
 var BHYVE_MIN_MEM_OVERHEAD = 256;
 
+exports.BHYVE_MEM_OVERHEAD = BHYVE_MEM_OVERHEAD;
+exports.BHYVE_MIN_MEM_OVERHEAD = BHYVE_MIN_MEM_OVERHEAD;
+
 // values used in feature table for KVM
 var KVM_MEM_OVERHEAD = 1024;
 var KVM_MIN_MEM_OVERHEAD = 256;
@@ -1493,7 +1495,7 @@ exports.properties = {
             min: 0
         },
         zfs: {
-            fields: ['quota'],
+            fields: ['refquota', 'quota'],
             types: ['filesystem']
         }
     }, ram: {
diff --git a/src/vm/node_modules/vmload/index.js b/src/vm/node_modules/vmload/index.js
index b5647728..fe2daeeb 100644
--- a/src/vm/node_modules/vmload/index.js
+++ b/src/vm/node_modules/vmload/index.js
@@ -309,10 +309,18 @@ function addRootDatasetProperties(vmobj, dataset_objects, options)
     }
 
     // dsobj.quota is in bytes, we want GiB for vmobj.quota
-    if (wantField(options, 'quota') && dsobj.hasOwnProperty('quota')) {
-        vmobj.quota = (dsobj.quota / (1024 * 1024 * 1024));
-        log.trace('found quota "' + vmobj.quota + '" for '
-            + vmobj.uuid);
+    if (vmobj.brand === 'bhyve') {
+        if (wantField(options, 'quota') && dsobj.hasOwnProperty('refquota')) {
+            vmobj.quota = (dsobj.refquota / (1024 * 1024 * 1024));
+            log.trace('found quota "' + vmobj.quota + '" for '
+                + vmobj.uuid);
+        }
+    } else {
+        if (wantField(options, 'quota') && dsobj.hasOwnProperty('quota')) {
+            vmobj.quota = (dsobj.quota / (1024 * 1024 * 1024));
+            log.trace('found quota "' + vmobj.quota + '" for '
+                + vmobj.uuid);
+        }
     }
 
     if (wantField(options, 'create_timestamp')
diff --git a/src/vm/node_modules/vmload/vmload-datasets.js b/src/vm/node_modules/vmload/vmload-datasets.js
index 6c987863..65cf249a 100644
--- a/src/vm/node_modules/vmload/vmload-datasets.js
+++ b/src/vm/node_modules/vmload/vmload-datasets.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  *
  */
 
@@ -168,7 +168,11 @@ function cleanDatasetObject(obj)
     // convert numeric fields to proper numbers
     number_fields.forEach(function (field) {
         if (obj.hasOwnProperty(field) && obj[field] !== '-') {
-            obj[field] = Number(obj[field]);
+            if (field === 'refquota' && obj[field] === 'none') {
+                obj[field] = 0;
+            } else {
+                obj[field] = Number(obj[field]);
+            }
         }
     });
 
diff --git a/src/vm/tests/test-vmload-datasets.js b/src/vm/tests/test-vmload-datasets.js
index 8b80e41a..f76537eb 100644
--- a/src/vm/tests/test-vmload-datasets.js
+++ b/src/vm/tests/test-vmload-datasets.js
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2014, Joyent, Inc. All rights reserved.
+ * Copyright (c) 2018, Joyent, Inc. All rights reserved.
  *
  */
 
@@ -71,6 +71,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -87,6 +88,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/01b2c898-945f-11e1-a523-af1afbe22822",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -103,6 +105,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/01b2c898-945f-11e1-a523-af1afbe22822@final",
             "quota": "-",
             "recsize": "-",
+            "refquota": "-",
             "refreservation": "-",
             "snapshot_limit": "-",
             "type": "snapshot",
@@ -119,6 +122,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25",
             "quota": 10737418240,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -135,6 +139,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25/data",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -151,6 +156,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/archive",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -167,6 +173,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/config",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -183,6 +190,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/cores",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -199,6 +207,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/cores/8be21e2a-ce25-4eb2-b796-b36b274b5a25",
             "quota": 107374182400,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -215,6 +224,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/cores/global",
             "quota": 10737418240,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -231,6 +241,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/dump",
             "quota": "-",
             "recsize": "-",
+            "refquota": "-",
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "volume",
@@ -247,6 +258,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/opt",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -263,6 +275,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/swap",
             "quota": "-",
             "recsize": "-",
+            "refquota": "-",
             "refreservation": 2147483648,
             "snapshot_limit": 18446744073709552000,
             "type": "volume",
@@ -279,6 +292,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/usbkey",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -295,6 +309,7 @@ test('test with one zone and one image', function (t) {
             "name": "zones/var",
             "quota": 0,
             "recsize": 131072,
+            "refquota": 0,
             "refreservation": 0,
             "snapshot_limit": 18446744073709552000,
             "type": "filesystem",
@@ -336,61 +351,62 @@ test('test with one zone and one image', function (t) {
      *
      * zfs list -H -p -t filesystem,snapshot,volume \
      *     -o compression,creation,filesystem_limit,mountpoint,name,quota,\
-     *     recsize,refreservation,snapshot_limit,type,userrefs,volblocksize,\
+     *     recsize,refquota,refreservation,snapshot_limit,type,userrefs,volblocksize,\
      *     volsize,zoned
      *
      * from a system with only one zone and one image.
      */
     var lines = [
         'off	1392846460	18446744073709552000	/zones	zones	0	'
-            + '131072	0	18446744073709552000	filesystem	-	-	-	'
-            + 'off',
+            + '131072	0	0	18446744073709552000	filesystem	-	-	'
+            + '-	off',
         'off	1392846722	18446744073709552000	'
             + '/zones/01b2c898-945f-11e1-a523-af1afbe22822	zones/'
-            + '01b2c898-945f-11e1-a523-af1afbe22822	0	131072	0	'
+            + '01b2c898-945f-11e1-a523-af1afbe22822	0	131072	none	0	'
             + '18446744073709552000	filesystem	-	-	-	off',
         '-	1335967116	-	-	zones/01b2c898-945f-11e1-a523-af1afbe22822@'
-            + 'final	-	-	-	-	snapshot	-	-	-	-',
+            + 'final	-	-	-	-	-	snapshot	-	-	-	-',
         'off	1392939536	18446744073709552000	'
             + '/zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25	zones/'
-            + '8be21e2a-ce25-4eb2-b796-b36b274b5a25	10737418240	131072	0	'
-            + '18446744073709552000	filesystem	-	-	-	off',
+            + '8be21e2a-ce25-4eb2-b796-b36b274b5a25	10737418240	131072	none'
+            + '	0	18446744073709552000	filesystem	-	-	-	off',
         'off	1392939536	18446744073709552000	'
             + '/zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25/'
             + 'data	zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25/data	0	'
-            + '131072	0	18446744073709552000	filesystem	-	-	-	on',
+            + '131072	none	0	18446744073709552000	filesystem	-	'
+            + '-	-	on',
         'lzjb	1392849677	18446744073709552000	/zones/archive	'
-            + 'zones/archive	0	131072	0	'
+            + 'zones/archive	0	131072	none	0	'
             + '18446744073709552000	filesystem	-	-	-	off',
         'off	1392846460	18446744073709552000	legacy	zones/config	'
-            + '0	131072	0	18446744073709552000	filesystem	'
+            + '0	131072	none	0	18446744073709552000	filesystem	'
             + '-	-	-	off',
         'lz4	1392846460	18446744073709552000	none	zones/cores	0	'
-            + '131072	0	18446744073709552000	filesystem	'
+            + '131072	none	0	18446744073709552000	filesystem	'
             + '-	-	-	off',
         'lz4	1392939536	18446744073709552000	'
             + '/zones/8be21e2a-ce25-4eb2-b796-b36b274b5a25/'
             + 'cores	zones/cores/8be21e2a-ce25-4eb2-b796-b36b274b5a25	'
-            + '107374182400	131072	0	18446744073709552000	filesystem	'
+            + '107374182400	131072	none	0	18446744073709552000	filesystem	'
             + '-	-	-	off',
         'lz4	1392846460	18446744073709552000	'
             + '/zones/global/cores	zones/cores/global	'
-            + '10737418240	131072	0	18446744073709552000	'
+            + '10737418240	131072	none	0	18446744073709552000	'
             + 'filesystem	-	-	-	off',
-        'off	1392846460	-	-	zones/dump	-	-	0	'
+        'off	1392846460	-	-	zones/dump	-	-	-	0	'
             + '18446744073709552000	volume	-	131072	'
             + '2146435072	-',
         'off	1392846460	18446744073709552000	legacy	zones/opt	0	'
-            + ' 131072	0	18446744073709552000	filesystem	'
+            + ' 131072	none	0	18446744073709552000	filesystem	'
             + '-	-	-	off',
-        'off	1392846461	-	-	zones/swap	-	-	2147483648	'
+        'off	1392846461	-	-	zones/swap	-	-	-	2147483648	'
             + '18446744073709552000	volume	'
             + '-	8192	4294967296	-',
         'off	1392846460	18446744073709552000	legacy	zones/usbkey	'
-            + '0	131072	0	'
+            + '0	131072	none	0	'
             + '18446744073709552000	filesystem	-	-	-	off',
         'off	1392846460	18446744073709552000	legacy	zones/var	0	'
-            + '131072	0	18446744073709552000	filesystem	'
+            + '131072	none	0	18446744073709552000	filesystem	'
             + '-	-	-	off'
     ];
 
@@ -402,13 +418,14 @@ test('test with one zone and one image', function (t) {
         'filesystem,snapshot,volume',
         '-o',
         'compression,creation,filesystem_limit,mountpoint,name,quota,'
-            + 'recsize,refreservation,snapshot_limit,type,userrefs,'
+            + 'recsize,refquota,refreservation,snapshot_limit,type,userrefs,'
             + 'volblocksize,volsize,zoned'
     ];
 
     var out = {};
 
     getDatasetsWrapper([], lines, out, function (err, dsobj) {
+        t.ifError(err, 'getDatasetsWrapper should have no error');
         t.equal(out.cmd, '/usr/sbin/zfs', 'zfs cmd is as expected');
         t.deepEqual(out.args, expected_args, 'zfs args are as expected');
         t.deepEqual(dsobj, expected_dsobj, 'dsobj parsed as expected');
